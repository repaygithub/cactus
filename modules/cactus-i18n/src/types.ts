import BaseI18nController, { LoadingState } from './BaseI18nController'

export interface ResourceDefinition {
  lang: string
  /**
   * FTL message resource
   */
  ftl: string
}

export interface I18nContextType {
  controller: BaseI18nController
  section: string
  lang: string
  loadingState: LoadingState
}
