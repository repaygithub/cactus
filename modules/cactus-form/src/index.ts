export { default as DependentField } from './DependentField'
export { default as FieldSpy } from './FieldSpy'
export { default as Field, validateRequired } from './Field'
