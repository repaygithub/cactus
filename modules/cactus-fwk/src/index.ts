export { useFeatureFlags, withFeatureFlags, FeatureFlag } from './featureFlags'
export { default } from './AppRoot'

export * from './types'
