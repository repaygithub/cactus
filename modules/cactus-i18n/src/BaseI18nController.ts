import 'intl'
import 'intl/locale-data/jsonp/en.js'
import 'intl/locale-data/jsonp/es.js'
import 'intl-pluralrules' // eslint-disable-line simple-import-sort/sort

import { FluentBundle, FluentFunction, FluentResource, FluentVariable } from '@fluent/bundle'
import { negotiateLanguages } from '@fluent/langneg'
import lodashSet from 'lodash/set'

interface HasI18nArgs {
  section: string
  id: string
  lang?: string
}

type SplitFunc = (key: string) => string | string[]
interface GetI18nArgs extends HasI18nArgs {
  args?: Record<string, FluentVariable>
  mapAttrs?: string | RegExp | SplitFunc
}

export interface I18nMessage {
  text: string | null
  attrs: Record<string, any>
  found: boolean
}

export type LoadState = 'new' | 'loading' | 'loaded' | 'error'
export type LoadEventType = 'loaded' | 'error' | 'all'
export type LoadEventHandler = (b: BundleInfo, prevState: LoadState, error?: any) => void

interface LoadEventListener {
  listener: LoadEventHandler
  eventType: LoadEventType
}

export type Loader = (
  this: BaseI18nController,
  bundleInfo: BundleInfo,
  loadOpts: LoadOpts
) => AsyncLoadResult

export interface LoadOpts extends Record<string, any> {
  loader?: Loader
  /** Call this instead of running through the list of global listeners. */
  onLoad?: LoadEventHandler
  // This is for hard-coded defaults, mostly intended for use by 3rd party libraries.
  defaults?: FluentResource
}

export interface LoadResult {
  resources: Resource[]
  version?: number
}

export type AsyncLoadResult = Promise<LoadResult | undefined>

export type Resource = BundleInfo | FluentResource

interface BundleOpts {
  /**
   * Flag to disable bidi isolating unicode characters.
   * @default = true
   */
  useIsolating: boolean

  /** Custom functions for Fluent to use during interpolation */
  functions: Record<string, FluentFunction>
}

interface I18nControllerOptions extends Partial<BundleOpts> {
  /**
   * Should be an array of locale codes
   */
  supportedLangs: string[]

  /**
   * fallback language if key does not exist
   */
  defaultLang?: string

  /** Set the current lang explicitly or uses provided default */
  lang?: string

  /** Flag for displaying error messages. Default is false */
  debugMode?: boolean
}

export default abstract class BaseI18nController {
  public lang = ''
  public defaultLang: string | undefined = undefined
  private _supportedLangs: string[]
  private _languages: string[] = []
  private _listeners: LoadEventListener[] = []
  private _bundles: Record<string, BundleInfo> = {}
  private _debugMode: boolean
  private _bundleOpts: BundleOpts

  public constructor({ functions = {}, ...options }: I18nControllerOptions) {
    if (!Array.isArray(options.supportedLangs) || options.supportedLangs.length === 0) {
      throw new Error('You must provide supported languages')
    } else if (options.defaultLang) {
      if (!options.supportedLangs.includes(options.defaultLang)) {
        throw new Error('The default language provided is not a supported language')
      }
      this.defaultLang = options.defaultLang
    } else if (!options.lang) {
      throw new Error('You must provide a `lang` or `defaultLang`')
    }
    if (this.load === undefined) {
      throw new Error('You must override the `load` method')
    }

    this._supportedLangs = options.supportedLangs
    this.setLang(options.lang || (options.defaultLang as string))
    this._debugMode = !!options.debugMode
    const useIsolating = options.useIsolating === undefined ? true : !!options.useIsolating
    this._bundleOpts = { useIsolating, functions }
  }

  protected abstract load(b: BundleInfo, l: LoadOpts): AsyncLoadResult

  protected _log(levelOrArg: string, ...args: any[]): void {
    if (this._debugMode) {
      if (console.hasOwnProperty(levelOrArg)) {
        // @ts-ignore
        console[levelOrArg](...args)
      } else {
        console.log(levelOrArg, ...args)
      }
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  protected getKey(id: string, section?: string, lang?: string): string {
    return id
  }

  protected negotiateLang(lang?: string, strict = false): string[] {
    if (lang === undefined) {
      return this._languages
    }
    const langs = negotiateLanguages([lang], this._supportedLangs, {
      defaultLocale: this.defaultLang,
      strategy: 'matching',
    })
    if (strict && !langs.length) {
      throw new Error(`Unsupported language: ${lang}`)
    }
    return langs
  }

  public setLang(lang: string): void {
    this._languages = this.negotiateLang(lang, true)
    this.lang = this._languages[0]
  }

  public get({
    args,
    section,
    id,
    lang: overrideLang = this.lang,
    mapAttrs,
  }: GetI18nArgs): I18nMessage {
    const langs = this.negotiateLang(overrideLang, true)
    const result: I18nMessage = { text: null, attrs: {}, found: false }
    for (const lang of langs) {
      const bundle = this.getBundle(section, lang)
      if (bundle) {
        const message = bundle.getMessage(this.getKey(id, section, lang))
        if (message) {
          result.found = true
          const { value, attributes } = message
          const errors: any[] = []
          if (value) {
            result.text = bundle.formatPattern(value, args, errors)
          }
          if (attributes) {
            for (const [attr, val] of Object.entries(attributes)) {
              const key =
                typeof mapAttrs === 'function'
                  ? mapAttrs(attr)
                  : mapAttrs
                  ? attr.split(mapAttrs)
                  : attr
              lodashSet(result.attrs, key, bundle.formatPattern(val, args, errors))
            }
          }
          if (errors.length) {
            this._log('warn', 'Errors while formatting translation', { id, section, lang, errors })
          }
          break
        } else {
          this._log('warn', 'Translation not found', { id, section, lang })
        }
      }
    }
    return result
  }

  public getText(args: GetI18nArgs): I18nMessage['text'] {
    return this.get(args).text
  }

  public hasText({ section, id, lang }: HasI18nArgs): boolean {
    lang = this.negotiateLang(lang)[0]
    const bundle = lang && this.getBundle(section, lang)
    return bundle ? bundle.hasMessage(this.getKey(id, section, lang)) : false
  }

  public _load(
    { lang: requestedLang, section }: { lang: string; section: string },
    loadOpts: LoadOpts = {}
  ): BundleInfo {
    const lang = this.negotiateLang(requestedLang, true)[0]
    const bundleInfo = this.getBundleInfo(section, lang, true) as BundleInfo
    if (bundleInfo.loadState === 'new' || bundleInfo.loadState === 'error') {
      const prevState = bundleInfo.loadState
      bundleInfo.loadState = 'loading'
      const loader = loadOpts.loader || this.load
      const onLoad = loadOpts.onLoad || this.onLoad
      loader.call(this, bundleInfo, loadOpts).then(
        (result) => {
          if (bundleInfo.update('loaded', loadOpts, result)) {
            onLoad.call(this, bundleInfo, prevState)
          }
        },
        (error) => {
          this._log('error', 'FTL Resource failed to load', { section, lang, error })
          bundleInfo.update('error', loadOpts)
          onLoad.call(this, bundleInfo, prevState, error)
        }
      )
    }
    return bundleInfo
  }

  public hasLoaded(section: string, lang?: string): boolean {
    const language = this.negotiateLang(lang)[0]
    return this.getLoadState(section, language) === 'loaded'
  }

  protected onLoad(bundleInfo: BundleInfo, prevState: LoadState, error?: unknown): void {
    for (const { listener, eventType } of this._listeners) {
      if (eventType === 'all' || eventType === bundleInfo.loadState) {
        listener(bundleInfo, prevState, error)
      }
    }
  }

  public addListener(listener: LoadEventHandler, eventType: LoadEventType = 'loaded'): void {
    this._listeners.push({ listener, eventType })
  }

  public removeListener(listener: LoadEventHandler): void {
    const index = this._listeners.findIndex((x) => x.listener === listener)
    if (index >= 0) {
      this._listeners.splice(index, 1)
    }
  }

  protected loadRef(
    referrer: BundleInfo,
    section: string,
    loadOpts: LoadOpts,
    lang?: string
  ): BundleInfo {
    const refBundle = this._load({ section, lang: lang || referrer.lang }, loadOpts)
    if (referrer === refBundle) {
      throw new Error(`Self-referencing section: ${section}/${refBundle.lang}`)
    } else if (referrer.lang !== refBundle.lang) {
      this._log('Ref to different language:', {
        section: referrer.section,
        lang: referrer.lang,
        ref: refBundle.section,
        refLang: refBundle.lang,
      })
    }
    refBundle.addDependent(referrer)
    return refBundle
  }

  public getLoadState(section: string, lang: string): LoadState {
    return this.getBundleInfo(section, lang)?.loadState || 'new'
  }

  protected getBundleInfo(section: string, lang: string, create = false): BundleInfo | undefined {
    const key = `${section}/${lang}`
    let bundleInfo = this._bundles[key]
    if (!bundleInfo && create) {
      this._bundles[key] = bundleInfo = new BundleInfo(section, lang)
    }
    return bundleInfo
  }

  protected getBundle(section: string, lang: string): FluentBundle | undefined {
    const bundleInfo = this.getBundleInfo(section, lang)
    if (!bundleInfo) {
      this._log('warn', 'Attempting to use unloaded section:', { section, lang })
      return
    }
    return bundleInfo.compile(this._bundleOpts)
  }
}

export class BundleInfo {
  public section: string
  public lang: string
  public loadState: LoadState = 'new'
  public loadOpts?: LoadOpts
  private bundle?: FluentBundle
  private messages?: FluentResource
  private version = 0
  private resources: Resource[] = []
  private dependents: Set<BundleInfo> = new Set()

  public constructor(section: string, lang: string) {
    this.section = section
    this.lang = lang
  }

  public addDependent(d: BundleInfo): void {
    this.dependents.add(d)
  }

  public compile(bundleOpts: BundleOpts): FluentBundle | undefined {
    if (this.bundle || this.loadState !== 'loaded') {
      return this.bundle
    }
    const resources: FluentResource[] = []
    for (const maybeRef of this.resources) {
      if (isRef(maybeRef)) {
        if (!maybeRef.compile(bundleOpts)) {
          // When it finishes loading it should mark this as dirty, leading to recompile.
          continue
        }
        resources.push(maybeRef.messages as FluentResource)
      } else {
        resources.push(maybeRef)
      }
    }
    const bundle = (this.bundle = new FluentBundle(this.lang, bundleOpts))
    for (const resource of resources) {
      bundle.addResource(resource, { allowOverrides: true })
    }
    this.messages = bundleToResource(bundle)
    this.version += 1
    return bundle
  }

  public update(loadState: LoadState, loadOpts: LoadOpts, result?: LoadResult): boolean {
    this.loadState = loadState
    this.loadOpts = loadOpts
    if (!result || this.resources === result.resources) return false
    else if (result.version) {
      if (result.version <= this.version) return false
      // We adjust here to get the right value after it's incremented on compile.
      this.version = result.version - 1
    }
    this.resources = result.resources
    this.markDirty()
    return true
  }

  public markDirty(): void {
    delete this.bundle
    this.dependents.forEach($markDirty)
  }
}
// I miss Python at times like this...
const $markDirty = (x: BundleInfo) => x.markDirty()

function isRef(maybeRef: Resource): maybeRef is BundleInfo
function isRef(maybeRef: any): boolean {
  return Array.isArray(maybeRef.resources)
}

type FluentPattern = FluentResource['body'][number]

function mapToResource(
  map: Map<string, FluentPattern>,
  resource: FluentResource = new FluentResource('')
) {
  map.forEach((entry) => resource.body.push(entry))
  return resource
}

function bundleToResource(bundle: FluentBundle) {
  const resource = new FluentResource('')
  mapToResource(bundle._terms, resource)
  mapToResource(bundle._messages, resource)
  return resource
}
