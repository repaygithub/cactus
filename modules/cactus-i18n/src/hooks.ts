import { createContext, useContext } from 'react'

import { I18nContextType } from './types'

export const I18nContext = createContext<I18nContextType | null>(null)

export const useI18nText = (id: string, args?: object, sectionOverride?: string) => {
  const context = useContext(I18nContext)
  if (context === null) {
    return null
  }
  const { controller, section } = context
  return controller.getText({ args, section: sectionOverride || section, id })
}

export const useI18nResource = (
  id: string,
  args?: object,
  sectionOverride?: string
): [string | null, object] => {
  const context = useContext(I18nContext)
  if (context === null) {
    return [null, {}]
  }
  const { controller, section } = context
  return controller.get({ args, section: sectionOverride || section, id })
}

export const useI18nContext = (section?: string) => {
  const context = useContext(I18nContext)
  if (context === null) {
    return null
  }
  const newContext = { ...context }
  if (section) {
    newContext.section = section
  }
  return newContext
}
