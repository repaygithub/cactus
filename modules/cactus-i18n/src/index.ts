export {
  I18nProvider as default,
  I18nSection,
  I18nElement,
  I18nFormatted,
  I18nResource,
  I18nText,
} from './Components'
export { useI18nText, useI18nResource, useI18nSection, I18nContext } from './hooks'
export { default as BaseI18nController } from './BaseI18nController'
export * from './types'
