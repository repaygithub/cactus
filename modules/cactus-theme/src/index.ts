import defaultTheme from './theme'
export * from './helpers/index'
export * from './theme'
export default defaultTheme
