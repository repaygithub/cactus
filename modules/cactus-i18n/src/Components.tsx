import { FluentVariable } from '@fluent/bundle'
import PropTypes from 'prop-types'
import React, {
  ComponentProps,
  JSXElementConstructor,
  useContext,
  useEffect,
  useReducer,
} from 'react'

import BaseI18nController from './BaseI18nController'
import { I18nContext, useI18nResource, useI18nText } from './hooks'
import { I18nContextType } from './types'

interface I18nProviderProps {
  children?: React.ReactNode
  controller: BaseI18nController
  lang?: string
  section?: string
}

const partialUpdate = (s: I18nContextType, a: Partial<I18nContextType>) => ({ ...s, ...a })
const initContext = (controller: BaseI18nController) => ({
  controller,
  lang: controller.lang,
  section: 'global',
})

const I18nProvider: React.FC<I18nProviderProps> = ({ section = 'global', ...props }) => {
  const controller = props.controller
  if (!(controller instanceof BaseI18nController)) {
    throw Error('I18nProvider must be given a controller which extends BaseI18nController')
  }
  if (typeof props.lang === 'string') {
    controller.setLang(props.lang)
  }
  const lang = controller.lang
  const [i18nContext, dispatch] = useReducer(partialUpdate, controller, initContext)
  if (
    lang !== i18nContext.lang ||
    section !== i18nContext.section ||
    controller !== i18nContext.controller
  ) {
    dispatch({ lang, section, controller })
  }
  useEffect(() => {
    if (controller instanceof BaseI18nController) {
      const triggerUpdate = () => dispatch({})
      controller.addListener(triggerUpdate)
      return (): void => controller.removeListener(triggerUpdate)
    }
  }, [controller])
  useEffect(() => {
    if (section) {
      controller.load({ section, lang })
    }
  }, [controller, section, lang])
  return <I18nContext.Provider value={i18nContext}>{props.children}</I18nContext.Provider>
}

I18nProvider.propTypes = {
  // TS doesn't like providing an abstract class below
  // @ts-ignore
  controller: PropTypes.instanceOf(BaseI18nController),
  lang: PropTypes.string,
  section: PropTypes.string,
}

interface DependencyLoadType {
  section: string
  [key: string]: any
}

interface I18nSectionProps {
  children?: React.ReactNode
  name: string
  lang?: string
  dependencies?: (string | DependencyLoadType)[]
  [key: string]: any
}

function hasLoadedAll(
  controller: BaseI18nController,
  section: string,
  lang?: string,
  dependencies?: (string | DependencyLoadType)[]
): boolean {
  if (!controller.hasLoaded(section, lang)) {
    return false
  }
  if (!Array.isArray(dependencies) || dependencies.length === 0) {
    return true
  }
  return dependencies.every((dep): boolean => {
    if (typeof dep === 'string') {
      return controller.hasLoaded(dep, lang)
    }
    return controller.hasLoaded(dep.section, lang)
  })
}

const I18nSection: React.FC<I18nSectionProps> = ({
  name: section,
  lang,
  dependencies,
  children,
  ...extraProps
}) => {
  const context = useContext(I18nContext)
  const ctxRef = React.useRef<I18nContextType | null>(null)
  if (context === null) {
    ctxRef.current = null
  } else {
    const { controller, lang: ctxLang } = context
    lang = controller.negotiateLang(lang || ctxLang, true)[0]
    // Don't change the context until everything is loaded.
    if (hasLoadedAll(controller, section, lang, dependencies)) {
      if (section === context.section && lang === ctxLang) {
        ctxRef.current = context
      } else {
        ctxRef.current = { controller, section, lang }
      }
    }
  }
  useEffect((): void => {
    if (lang && context !== null) {
      const { controller } = context
      controller.load({ lang, section }, extraProps)
      if (Array.isArray(dependencies)) {
        for (const dep of dependencies) {
          if (!dep) continue
          else if (typeof dep === 'string') {
            controller.load({ lang, section: dep })
          } else {
            const { section: depSection, ...depExtra } = dep
            controller.load({ lang, section: depSection }, depExtra)
          }
        }
      }
    }
  })
  // wait to render until section is loaded
  if (context !== null && ctxRef.current === null) {
    return null
  }
  return <I18nContext.Provider value={ctxRef.current}>{children}</I18nContext.Provider>
}

I18nSection.propTypes = {
  name: PropTypes.string.isRequired,
  lang: PropTypes.string,
}

interface I18nTextProps {
  get: string
  args?: Record<string, FluentVariable>
  section?: string
}

const I18nText: React.FC<React.PropsWithChildren<I18nTextProps>> = (props) => {
  const text = useI18nText(props.get, props.args, props.section)
  return <React.Fragment>{text || props.children || props.get}</React.Fragment>
}

I18nText.propTypes = {
  get: PropTypes.string.isRequired,
  args: PropTypes.object as PropTypes.Validator<Record<string, FluentVariable>>,
  section: PropTypes.string,
}

type TagNameOrReactComp = keyof JSX.IntrinsicElements | JSXElementConstructor<any>

type I18nElementProps<Elem extends TagNameOrReactComp> = {
  as: Elem
} & Partial<ComponentProps<Elem>> &
  I18nTextProps

const I18nElement = function <Elem extends TagNameOrReactComp>(
  props: I18nElementProps<Elem>
): React.ReactElement {
  const { get, args = {}, section, as, ...rest } = props
  const { text, attrs } = useI18nResource(get, args, section)
  const elemProps = { ...rest, ...attrs }
  return React.createElement(as, elemProps, text || get)
}

I18nElement.propTypes = {
  get: PropTypes.string.isRequired,
  args: PropTypes.object,
  section: PropTypes.string,
  as: PropTypes.elementType,
}

interface I18nResourceProps extends I18nTextProps {
  render?: (message: string, attributes?: { [k: string]: any } | null) => React.ReactNode
  children?: (message: string, attributes?: { [k: string]: any } | null) => React.ReactNode
}

const I18nResource: React.FC<I18nResourceProps> = (props) => {
  const { text, attrs } = useI18nResource(props.get, props.args, props.section)
  let renderer = null
  if (typeof props.children === 'function') {
    renderer = props.children
  } else if (typeof props.render === 'function') {
    renderer = props.render
  }

  return (
    <React.Fragment>{renderer !== null ? renderer(text || props.get, attrs) : null}</React.Fragment>
  )
}

I18nResource.propTypes = {
  get: PropTypes.string.isRequired,
  args: PropTypes.object as PropTypes.Validator<Record<string, FluentVariable>>,
  section: PropTypes.string,
  render: PropTypes.func,
  children: PropTypes.func,
}

interface I18nFormattedProps extends I18nTextProps {
  formatter: (message: string) => React.ReactNode
}

const I18nFormatted: React.FC<I18nFormattedProps> = (props) => {
  const text = useI18nText(props.get, props.args, props.section)
  return <React.Fragment>{text !== null ? props.formatter(text) : props.get}</React.Fragment>
}

I18nFormatted.propTypes = {
  get: PropTypes.string.isRequired,
  args: PropTypes.object as PropTypes.Validator<Record<string, FluentVariable>>,
  section: PropTypes.string,
  formatter: PropTypes.func.isRequired,
}

export { I18nProvider, I18nSection, I18nElement, I18nFormatted, I18nResource, I18nText }
