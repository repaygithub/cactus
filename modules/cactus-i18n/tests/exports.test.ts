import * as cactusI18n from '../src/index'

describe('@repay/cactus-i18n', () => {
  test('should match snapshot of exports', () => {
    expect(Object.keys(cactusI18n)).toMatchSnapshot()
  })
})
