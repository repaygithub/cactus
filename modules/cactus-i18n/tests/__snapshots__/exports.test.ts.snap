// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@repay/cactus-i18n should match snapshot of exports 1`] = `
Array [
  "BaseI18nController",
  "I18nContext",
  "I18nElement",
  "I18nFormatted",
  "I18nResource",
  "I18nSection",
  "I18nText",
  "default",
  "useI18nResource",
  "useI18nText",
  "BundleInfo",
]
`;
